Fix StartupOrderResolver timing issue

jsdjayanga opened the pull request:
Resolves #1244

jsdjayanga committed:
Fix StartupOrderResolver timing issue

This commit fixes the timing issue had in the StartupOrderResolver. Introduced a new StartupServiceCache to track the OSGi services received by the components and changed the pending capability checking logic accordingly.

jsdjayanga committed:
Rename method updateIndirectCapability to updateCapability

jsdjayanga committed:
Fix pending capability identification code

jsdjayanga committed:
Remove unwanted comments and renaming some variables

jsdjayanga committed:
Add licence header to the new files

jsdjayanga committed:
Add more comments

jsdjayanga committed:
Update ResolvingtheComponentStartupOrder.md

Added a new subsection "Notifying the StartupServiceCache about received services"

jsdjayanga committed:
Fix some synchronization issues

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/runtime/RuntimeServiceListenerComponent.java:
```diff
@@ -70,6 +71,7 @@ protected void start(BundleContext bundleContext) {
     protected void registerRuntime(Runtime runtime) {
         try {
             runtimeManager.registerRuntime(runtime);
+            StartupServiceUtils.updateServiceCache("carbon-runtime-mgt", Runtime.class, runtime);
```
Let's use a constant for the string value on both places.

jsdjayanga replied:
Updated the component names to constants via 00640b0b427de47168b56103e30a5709b419cdb1

jsdjayanga committed:
Add constants for component names

Kishanthan commented on docs/KernelFeatures/ResolvingtheComponentStartupOrder.md:
```diff
@@ -103,6 +104,25 @@ An OSGi listener component is defined as shown below.
+          )
+          protected void registerTransport(CarbonTransport transport, Map<String, ?> ref) {
+              transportManager.registerTransport(transport);
+              StartupServiceUtils.updateServiceCache("carbon-transport-mgt", CarbonTransport.class, transport);
```
Update description about what is the "componentName" used here

jsdjayanga replied:
Updated the componentName description
1672758b2aef709f21706a89d19b4e28f5a4c9ad

Kishanthan replied:
Above doc is not updated with the recent changes to method arguments

jsdjayanga committed:
Update ResolvingtheComponentStartupOrder.md

Explained more about componentName

jsdjayanga committed:
Capitalize component annotation

jsdjayanga committed:
Explain the usage of method parameters in StartupServiceUtils

jsdjayanga committed:
Remove serviceInstance parameter from the StartupServiceUtils

Removed the serviceInstance parameter from the StartupServiceUtils and StartupServiceCache. Change the StartupServiceCache.getAvailableService() accordingly.

jsdjayanga committed:
Modify ResolvingtheComponentStartupOrder.md

Updated "Notifying the StartupServiceCache about received services" section acccording to the new API change.

jsdjayanga committed:
Merge branch 'master' into startup-order-resolver9

jsdjayanga committed:
Fix merge conflicts

jsdjayanga committed:
Add test case for StartupOrderResolver fix

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/startupresolver/beans/StartupComponent.java:
```diff
@@ -118,24 +123,62 @@ public boolean isServiceRequired(String service) {
+        synchronized (expectedCapabilityList) {
+            if (capability.getState() == Capability.CapabilityState.EXPECTED) {
+                Optional<Capability> optCapability = expectedCapabilityList.stream()
+                        .filter(cap -> cap.getName().equals(capability.getName()))
```
Lets use "capability" instead of "cap" for better readability.

jsdjayanga replied:
There is already a variable with name capability. That is why it was used as cap

jsdjayanga replied:
Updated the variable name to expCapability.
dd6a3eae5a4d5994bc1f5c78582bd06c45fb4382

Kishanthan commented on core/src/main/java/org/wso2/carbon/kernel/internal/startupresolver/beans/Capability.java:
```diff
@@ -49,11 +51,21 @@
+                capability.getType(),
+                capability.getState(),
+                capability.getBundle(),
+                capability.isDirectDependency());
```
Can't we keep everything in a single line here.

jsdjayanga replied:
Done via dd6a3eae5a4d5994bc1f5c78582bd06c45fb4382

jsdjayanga committed:
Rename some variables and Formatting

jsdjayanga committed:
Rename the Custom3 test artifact to file and Adding doc comments for Capability
